File: JavaBackend/src/main/java/SUITS2025Backend/PoiList/PoiController.java
package SUITS2025Backend.PoiList;

import SUITS2025Backend.db.PoiDbController;
import SUITS2025Backend.db.PoiResponse;
import io.javalin.Javalin;
import io.javalin.http.Context;

import java.util.List;

public class PoiController {
    private static final PoiDbController poiDbController = new PoiDbController();

    public static void setup(Javalin app) {
        app.get("/poi", PoiController::getPois);
        app.post("/poi", PoiController::addPoi);
        app.delete("/poi/{name}", PoiController::removePoi);
        app.get("/poi/tag/{tag}", PoiController::getPoisByTag);
        app.get("/poi/area", PoiController::getPoisInArea);
        app.post("/audio", poiDbController::submitAudio);
        app.get("/audio/{id}", poiDbController::getAudio);
        app.delete("/poi", poiDbController::deletePois);
    }

    private static void getPois(Context ctx) {
        List<PoiResponse> resp = poiDbController.getPois();
        System.out.println(resp);
        ctx.json(resp);
    }

    private static void addPoi(Context ctx) {
        PoiResponse poi = ctx.bodyAsClass(PoiResponse.class);
        poiDbController.addPoi(poi);
        ctx.result("Successfully added POI");
    }

    private static void removePoi(Context ctx) {
        String name = ctx.pathParam("name");
        poiDbController.deletePoi(name);
        ctx.result("Successfully deleted POI");
    }

    private static void getPoisByTag(Context ctx) {
        String tag = ctx.pathParam("tag");
        List<PoiResponse> resp = poiDbController.getPoisByTag(tag);
        ctx.json(resp);
    }

    private static void getPoisInArea(Context ctx) {
        double minLat = Double.parseDouble(ctx.queryParam("minLat"));
        double maxLat = Double.parseDouble(ctx.queryParam("maxLat"));
        double minLon = Double.parseDouble(ctx.queryParam("minLon"));
        double maxLon = Double.parseDouble(ctx.queryParam("maxLon"));
        
        List<PoiResponse> resp = poiDbController.getPoisInArea(minLat, maxLat, minLon, maxLon);
        ctx.json(resp);
    }
    
}

File: JavaBackend/src/main/java/SUITS2025Backend/PoiList/PoiController.kt
package SUITS2025Backend.PoiList

import SUITS2025Backend.db.PoiDbController
import SUITS2025Backend.db.PoiResponse
import SUITS2025Backend.db.PoiResponseJson
import io.javalin.Javalin
import io.javalin.http.Context

/** Singleton controller that wires POI–related routes */
object PoiController {

    private val poiDbController = PoiDbController()

    /** Call from your Server.kt / main to register routes */
    fun setup(app: Javalin) = with(app) {

        // --- POI core -------------------------------------------------------
        get("/poi")                 { ctx -> getPois(ctx) }
        post("/poi")                { ctx -> addPoi(ctx) }
        put("/poi")                 { ctx -> updatePoi(ctx) }

        // --- Tags & areas ---------------------------------------------------
        get("/poi/tag/{tag}")       { ctx -> getPoisByTag(ctx) }
        get("/poi/area")            { ctx -> getPoisInArea(ctx) }
        post("/poi/updateTags/{id}"){ ctx -> poiDbController.updateTags(ctx) }

        // --- Audio & voice notes -------------------------------------------
        post("/audio")              { ctx -> poiDbController.submitAudio(ctx) }
        get("/audio/{id}")          { ctx -> poiDbController.getAudio(ctx) }
        post("/poi/addVoiceNote")   { ctx -> poiDbController.addVoiceNote(ctx) }

        // --- Delete helpers -------------------------------------------------
        delete("/poi")              { ctx -> poiDbController.deletePois(ctx) }
        delete("/poi/{id}")         { ctx -> poiDbController.deletePoi(ctx) }
    }

    // -----------------------------------------------------------------------
    // Handlers that delegate to the DB layer
    // -----------------------------------------------------------------------

    private fun getPois(ctx: Context) {
        val resp: PoiResponseJson = poiDbController.getPois()

        ctx.json(resp)
    }

    private fun addPoi(ctx: Context) {
        val poi = ctx.bodyAsClass(PoiResponse::class.java)
        val resp = poiDbController.addPoi(poi)
        ctx.json(resp)
    }



    private fun updatePoi(ctx: Context) {
        poiDbController.updatePoi(ctx)
    }

    private fun getPoisByTag(ctx: Context) {
        val tag = ctx.pathParam("tag")
        val resp = poiDbController.getPoisByTag(tag)
        ctx.json(resp)
    }

    private fun getPoisInArea(ctx: Context) {
        val minLat  = ctx.queryParam("minLat")!!.toDouble()
        val maxLat  = ctx.queryParam("maxLat")!!.toDouble()
        val minLon  = ctx.queryParam("minLon")!!.toDouble()
        val maxLon  = ctx.queryParam("maxLon")!!.toDouble()

        val resp = poiDbController.getPoisInArea(minLat, maxLat, minLon, maxLon)
        ctx.json(resp)
    }
}
